File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.cc
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define X first
#define Y second

int main(void) {
    ios::sync_with_stdio(0), cin.tie(0);
    
    while(true) {
        int n;
        cin>>n;
        if(n==0) break; // if input value is zero, stop loop
        
        stack<pair<ll, ll>> S;
        ll ans = 0;
        
        for(int i=0;i<n;i++) {
            int h;
            cin>>h;
            int idx = i; // this value is square's width
            
            while(!S.empty() && S.top().X >= h) {
                ans = max(ans, (i-S.top().Y) * S.top().X); // if stack's top height is higher than input value
                // Calculate the area of all of the rectangles which we can make and compare them
                idx = S.top().Y; // Update the area that we can create with the current height of the rectangle.
                S.pop();
            }
            S.push({h, idx});
        }
        while(!S.empty()) {
            ans = max(ans, (n-S.top().Y) * S.top().X);
            S.pop();
        }
        cout<<ans<<'\n';
    }
    
    return 0;
}

File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.py
import sys
r = sys.stdin.readline
sys.setrecursionlimit(10**6)

def solution():
    while True:
        hl = list(map(int, r().split()))
        n = hl.pop(0)
        
        if not n: break
        S = []
        ans = 0
        
        for i in range(n):
            idx  = i
            while S and S[-1][0] >= hl[i]:
                ans = max(ans, (i-S[-1][1])*S[-1][0])
                idx = S[-1][1]
                S.pop()
            S.append((hl[i], idx))
            
        while S:
            ans = max(ans, (n-S[-1][1])*S[-1][0])
            S.pop()
        print(ans)
    
solution()